ARCs/arc-0020.md

```diff
@@ -0,0 +1,869 @@
+  }
+]
+```
+
+Calling `asset_freeze` prevents any transfer of a Smart ASA. Calling
+`account_freeze` prevents a specific account from transferring or receiving a
+Smart ASA.
+
+> Upon a call to `asset_freeze` or `account_freeze`, a reference implementation
+> SHOULD:
+>
+> - Fail if `freeze_asset` does not correspond to an ASA controlled by this smart
+>   contract.
+> - Succeed iff the `sender` of the transaction corresponds to the `freeze_addr`,
+>   as persisted by the controlling Smart Contract.
+>
+> In addition:
+>
+> - Upon a call to `asset_freeze`, the controlling Smart Contract SHOULD persist
+>   the tuple `(freeze_asset, asset_frozen)` (for instance, by setting a `frozen`
+>   flag in _global_ storage).
+> - Upon a call to `account_freeze` the controlling Smart Contract SHOULD persist
+>   the tuple `(freeze_asset, freeze_account, asset_frozen)` (for instance by
+>   setting a `frozen` flag in the _local_ storage of the `freeze_account`). See
+>   the [security considerations section](#security-considerations) for how to
+>   ensure that Smart ASA holders cannot reset their `frozen` flag by clearing out
+>   their state at the controlling Smart Contract.
+
+```json
+[
+  {
+    "name": "is_asset_frozen",
+    "args": [{ "type": "asset", "name": "freeze_asset" }],
+    "returns": { "type": "bool" }
+  },
+  {
+    "name": "is_account_frozen",
+    "args": [
+      { "type": "asset", "name": "freeze_asset" },
+      { "type": "account", "name": "freeze_account" }
+    ],
+    "returns": { "type": "bool" }
+  }
+]
+```
+
+The value return by `is_asset_frozen` (respectively, `is_account_frozen`) tells
+whether any account (respectively `freeze_account`) can transfer or receive
+`freeze_asset`. A `false` value indicates that the transfer will be rejected.
+
+> Upon a call to `is_asset_frozen`, a reference implementation SHOULD retrieve
+> the tuple `(freeze_asset, asset_frozen)` as stored on `asset_freeze` and
+> return the value corresponding to `asset_frozen`.
+>
+> Upon a call to `is_account_frozen`, a reference implementation SHOULD retrieve the
+> tuple `(freeze_asset, freeze_account, asset_frozen)` as stored on `account_freeze`
+> and return the value corresponding to `asset_frozen`.
+
+#### Asset Destroy
+
+```json
+{
+  "name": "asset_destroy",
+  "args": [{ "type": "asset", "name": "destroy_asset" }],
+  "returns": { "type": "void" }
+}
+```
+
+Calling `asset_destroy` destroys a Smart ASA.
+
+> Upon a call to `asset_destroy`, a reference implementation SHOULD:
+>
+> - Fail if `destroy_asset` does not correspond to an ASA controlled by this smart
+>   contract.
+>
+> It is RECOMMENDED for calls to this method to be permissioned (see
+> `asset_create`).
+>
+> The controlling Smart Contract SHOULD perform an asset destroy operation on
+> the ASA with ID `destroy_asset`. The operation will fail if the asset is still
+> in circulation.
+
+#### Circulating Supply
+
+```json
+{
+  "name": "circulating_supply",
```

Comment:
Should we have a naming convention for read-only methods that they start with `get_*`?

Reply:
Other `get_*` methods might do math as well, so I'd be ok with renaming this `get_circulating_supply`.

Reply:
I fine with standardise `get_` naming convention.

Reply:
+1 on `get_`